File: js/src/swmmrs/objects/amm.ts
import type { Call } from "../protocol.js";

/** Standard runoff-response or seasonal baseflow AMM component. */
export type AmmComponentKind = "standard" | "baseflow";

/** Detached standard AMM component. `kind` is exactly `"standard"`. */
export interface AmmStandardComponent {
  /** Standard response discriminator. */
  readonly kind: "standard";
  /** Component identity within this model. */
  readonly id: string;
  /** Dry-weather capture parameter in the model's native capture units. */
  readonly dryCapture: number;
  /** Precipitation averaging window in hours. */
  readonly precipitationWindowHours: number;
  /** Response hydrograph half-life in hours. */
  readonly hydrographHalfLifeHours: number;
  /** Initial wet capture parameter in native capture units. */
  readonly initialWetCapture: number;
  /** Antecedent moisture half-life in hours. */
  readonly moistureHalfLifeHours: number;
  /** Temperature averaging window in hours. */
  readonly temperatureWindowHours: number;
  /** Spring cold seasonal hydrograph capture factor. */
  readonly springColdShcf: number;
  /** Hot seasonal hydrograph capture factor. */
  readonly hotShcf: number;
  /** Fall cold seasonal hydrograph capture factor. */
  readonly fallColdShcf: number;
}

/** Detached baseflow AMM component. `kind` is exactly `"baseflow"`. */
export interface AmmBaseflowComponent {
  /** Seasonal baseflow discriminator. */
  readonly kind: "baseflow";
  /** Component identity within this model. */
  readonly id: string;
  /** Precipitation averaging window in hours. */
  readonly precipitationWindowHours: number;
  /** Response hydrograph half-life in hours. */
  readonly hydrographHalfLifeHours: number;
  /** Temperature averaging window in hours. */
  readonly temperatureWindowHours: number;
  /** Spring cold capture parameter in native capture units. */
  readonly springColdCapture: number;
  /** Hot capture parameter in native capture units. */
  readonly hotCapture: number;
  /** Fall cold capture parameter in native capture units. */
  readonly fallColdCapture: number;
}

/** AMM component discriminated by `kind`; component arrays are replaced as a whole. */
export type AmmComponent = AmmStandardComponent | AmmBaseflowComponent;

/** Complete AMM model configuration in native project units. */
export interface AmmModelConfiguration {
  /** Canonical rain-gage ID. */
  readonly rainGage: string;
  /** Cold reference temperature in the model's configured temperature units. */
  readonly coldTemperature: number;
  /** Hot reference temperature in the model's configured temperature units. */
  readonly hotTemperature: number;
  /** Complete ordered standard/baseflow component list. */
  readonly components: readonly AmmComponent[];
}

/** Sparse AMM model update. Omitted fields retain their current values. */
export interface AmmModelPatch {
  /** {@inheritDoc AmmModelConfiguration.rainGage} */
  rainGage?: string;
  /** {@inheritDoc AmmModelConfiguration.coldTemperature} */
  coldTemperature?: number;
  /** {@inheritDoc AmmModelConfiguration.hotTemperature} */
  hotTemperature?: number;
  /** Replacement complete component array; omission retains it. */
  components?: readonly AmmComponent[];
}

/** Detached AMM node assignment in native project units. IDs are canonical. */
export interface AmmAssignment {
  /** Canonical receiving node ID. */
  readonly node: string;
  /** Canonical AMM-model ID. */
  readonly model: string;
  /** Assigned land area in acres (US) or hectares (SI). */
  readonly area: number;
}

export interface AmmOperations {
  ammModelConfiguration: { args: [string]; result: AmmModelConfiguration };
  configureAmmModel: { args: [string, AmmModelPatch]; result: void };
  ammAssignments: { args: []; result: readonly AmmAssignment[] };
  replaceAmmAssignments: { args: [readonly AmmAssignment[]]; result: void };
}

/** Editable AMM definition from `simulation.ammModels`; node assignments are managed on the simulation. */
export class AmmModel {
  /** Canonical configured model ID. */
  readonly id: string;
  readonly #call: Call;
  private constructor(id: string, call: Call) { this.id = id; this.#call = call; Object.freeze(this); }
  /** @internal */
  static create(id: string, call: Call): AmmModel { return new AmmModel(id, call); }
  /** Read the definition in any healthy owner state.
   * @returns Detached complete AMM configuration.
   */
  configuration(): Promise<AmmModelConfiguration> { return this.#call("ammModelConfiguration", this.id); }
  /** Atomically update an AMM definition in `open` or `ended`; accepted edits return the owner to `open`.
   * @param patch - Sparse fields; supplying components replaces the complete array.
   * @returns Resolves after all fields are accepted.
   */
  configure(patch: AmmModelPatch): Promise<void> { return this.#call("configureAmmModel", this.id, patch); }
}

File: python/src/swmmrs/objects/_amm.py
"""Typed editable AMM model views."""

from __future__ import annotations as _annotations

from collections.abc import Iterable as _Iterable
from typing import TYPE_CHECKING as _TYPE_CHECKING, cast as _cast, final as _final

from .._prettier import pretty_dataclass as _pretty_dataclass
from ._base import _LiveView
from ._subcatchments import RainGage

if _TYPE_CHECKING:
    from ._nodes import Node


@_final
@_pretty_dataclass(slots=True)
class AmmStandardComponent:
    """Store one detached standard AMM component in project units."""

    id: str
    dry_capture: float
    precipitation_window_hours: float
    hydrograph_half_life_hours: float
    initial_wet_capture: float
    moisture_half_life_hours: float
    temperature_window_hours: float
    spring_cold_shcf: float
    hot_shcf: float
    fall_cold_shcf: float


@_final
@_pretty_dataclass(slots=True)
class AmmBaseflowComponent:
    """Store one detached baseflow AMM component in project units."""

    id: str
    precipitation_window_hours: float
    hydrograph_half_life_hours: float
    temperature_window_hours: float
    spring_cold_capture: float
    hot_capture: float
    fall_cold_capture: float


AmmComponent = AmmStandardComponent | AmmBaseflowComponent


@_final
class AmmModel(_LiveView):
    """Expose a generation-bound editable AMM model."""

    __slots__ = ()

    def _configuration(
        self,
    ) -> tuple[RainGage, float, float, tuple[tuple[object, ...], ...]]:
        return _cast(
            tuple[RainGage, float, float, tuple[tuple[object, ...], ...]],
            self._simulation._owner.amm_model_configuration(
                self._simulation,
                self._generation,
                self._index,
                self._id,
                self._subtype,
            ),
        )

    def update(self, **changes: object) -> None:
        """Atomically update supplied AMM model fields in project units."""
        if "components" in changes:
            components = changes["components"]
            if not isinstance(components, _Iterable):
                raise TypeError("components must be an iterable of AMM components")
            changes["components"] = tuple(
                _component_to_native(component) for component in components
            )
        self._simulation._owner.update_amm_model(
            self._generation, self._index, self._id, self._subtype, changes
        )

    @property
    def rain_gage(self) -> RainGage:
        """Return the model's rain gage."""
        return self._configuration()[0]

    @rain_gage.setter
    def rain_gage(self, value: RainGage) -> None:
        self.update(rain_gage=value)

    @property
    def cold_temperature(self) -> float:
        """Return the cold reference temperature in project units."""
        return self._configuration()[1]

    @cold_temperature.setter
    def cold_temperature(self, value: float) -> None:
        self.update(cold_temperature=value)

    @property
    def hot_temperature(self) -> float:
        """Return the hot reference temperature in project units."""
        return self._configuration()[2]

    @hot_temperature.setter
    def hot_temperature(self, value: float) -> None:
        self.update(hot_temperature=value)

    @property
    def components(self) -> tuple[AmmComponent, ...]:
        """Return detached component definitions in configured order."""
        return tuple(_component_from_native(component) for component in self._configuration()[3])

    @components.setter
    def components(self, value: _Iterable[AmmComponent]) -> None:
        self.update(components=value)


@_final
@_pretty_dataclass(slots=True)
class AmmAssignment:
    """Store one detached AMM node assignment in project units."""

    node: Node
    model: AmmModel
    area: float


def _component_from_native(value: tuple[object, ...]) -> AmmComponent:
    kind = value[0]
    if kind == "standard":
        (
            _,
            identifier,
            dry_capture,
            precipitation_window_hours,
            hydrograph_half_life_hours,
            initial_wet_capture,
            moisture_half_life_hours,
            temperature_window_hours,
            spring_cold_shcf,
            hot_shcf,
            fall_cold_shcf,
        ) = _cast(
            tuple[str, str, float, float, float, float, float, float, float, float, float],
            value,
        )
        return AmmStandardComponent(
            identifier,
            dry_capture,
            precipitation_window_hours,
            hydrograph_half_life_hours,
            initial_wet_capture,
            moisture_half_life_hours,
            temperature_window_hours,
            spring_cold_shcf,
            hot_shcf,
            fall_cold_shcf,
        )
    if kind == "baseflow":
        (
            _,
            identifier,
            precipitation_window_hours,
            hydrograph_half_life_hours,
            temperature_window_hours,
            spring_cold_capture,
            hot_capture,
            fall_cold_capture,
        ) = _cast(tuple[str, str, float, float, float, float, float, float], value)
        return AmmBaseflowComponent(
            identifier,
            precipitation_window_hours,
            hydrograph_half_life_hours,
            temperature_window_hours,
            spring_cold_capture,
            hot_capture,
            fall_cold_capture,
        )
    raise RuntimeError(f"unknown native AMM component kind: {kind!r}")


def _component_to_native(component: object) -> tuple[object, ...]:
    if isinstance(component, AmmStandardComponent):
        return (
            "standard",
            component.id,
            component.dry_capture,
            component.precipitation_window_hours,
            component.hydrograph_half_life_hours,
            component.initial_wet_capture,
            component.moisture_half_life_hours,
            component.temperature_window_hours,
            component.spring_cold_shcf,
            component.hot_shcf,
            component.fall_cold_shcf,
        )
    if isinstance(component, AmmBaseflowComponent):
        return (
            "baseflow",
            component.id,
            component.precipitation_window_hours,
            component.hydrograph_half_life_hours,
            component.temperature_window_hours,
            component.spring_cold_capture,
            component.hot_capture,
            component.fall_cold_capture,
        )
    raise TypeError("components must contain AmmStandardComponent or AmmBaseflowComponent")
